modules/api/pkg/resources/machine/machinedeployment.go

```diff
@@ -316,26 +316,13 @@ func getProviderConfig(c *kubermaticv1.Cluster, nd *apiv1.NodeDeployment, dc *ku
 	return &config, nil
 }
 
-func getOperatingSystemProfile(nd *apiv1.NodeDeployment, dc *kubermaticv1.Datacenter) string {
-	if dc.Spec.DefaultOperatingSystemProfiles == nil {
-		return ""
+// getOperatingSystemProfile returns the OSP selected while creating the cluster.
+func getOperatingSystemProfile(nd *apiv1.NodeDeployment) string {
+	if osp := nd.Annotations[osmresources.MachineDeploymentOSPAnnotation]; osp != "" {
+		return osp
 	}
```

Comment:
OSP is an optional value(passed in annotation) here and it's not required for frontend or API to always pass it. In case the value is not provided, there are two options:
1. When defaults are defined at datacenter level, use them.
2. Otherwise set "k8c.io/operating-system-profile" as empty string and let OSM default/mutate the value.

We can't remove the datacenter default handling from here.

Reply:
Thanks @ahmedwaleedmalik for the detailed explanation! Really appreciated it.

I've updated the code accordingly, please let me know if it needs more work. Thanks!